### PR TITLE
[#89017762] First step to adding consistent authorization

### PR DESCRIPTION
**Pull request by cmc333333**
Opening this PR to see what we think of this direction. Using Pundit, policy checking for restarting a cart is now centralized.

Added a mechanism for combining policy checks, as a single "can this user edit?" could fail due for multiple reasons.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,10 +1,34 @@
...
+    else
+      queries = [query]
+    end
+
+    queries.each do |q|
+      super(record, q)  # default processing
+    end
+
+    true
+  end
```
Comment (afeld) [inline app/controllers/application_controller.rb]: Mind adding some inline comments to explain what's going on in more detail?

File: app/policies/proposal_policy.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    }
+  end
+
+  def initialize(user, proposal)
+    @user = user
+    @proposal = proposal
+  end
+
+  def author?
+    @proposal.requester_id == @user.id
```
Comment (afeld) [inline app/policies/proposal_policy.rb]: Minor: if you're using `attr_reader`, access the attributes with `self.ATTRNAME`. If you spell instance variables wrong, they give back `nil`, which is really annoying, so I tend to use them as little as possible.

File: app/policies/tree_policy.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    else    # deferring to another query/permission
+      self.flatten_tree(value)
+    end
+  end
+
+  def test_all(key)
+    values = self.flatten_tree(key).map{ |k| self.send(k) }
+    values.all?
+  end
+end
```
Comment (afeld) [inline app/policies/tree_policy.rb]: This seems like it could make a great wrapper gem for pundit! Or maybe a PR to the original.

**Comment by afeld** [comment]
:heart_eyes: 

If this works, I say :shipit: Happy to deal with authorization for different parts of the application in different pieces.

File: app/policies/proposal_policy.rb (hunk 2 above)
Comment (cmc333333) [inline app/policies/proposal_policy.rb]: The above diff is misleading; `attr_reader`'s been removed now

File: app/policies/tree_policy.rb (hunk 3 above)
Comment (cmc333333) [inline app/policies/tree_policy.rb]: Mayhaps. Let's use pundit a bit more -- we very well might be using it "wrong"

**Comment by cmc333333** [comment]
Updated; I think it can be merged before all of the policies are in place.

**Pull request merged**